### PR TITLE
feat(web): prototype KMX+ TouchLayout generator (in TS)

### PR DESCRIPTION
**Pull request by jahorton**
In order to build the touch-layout required for OSK construction for KMX+ keyboards, we either need that layout to be pre-compiled... or we need the ability to parse a KMX+ file in order to find the data we need.  This PR represents initial efforts toward the latter - parsing the KMX+ format in order to construct a viable TouchLayout that the OSK may consume.

<details><summary>Actual converted layout!</summary>

Using the KMX+ keyboard `imperial_aramaic.kmx`, as output for https://github.com/keymanapp/keyboards/pull/2993...

```json
[
  {
    "id": "default",
    "row": [
      {
        "id": 0,
        "key": [
          {
            "id": "K_BKQUOTE",
            "text": "`",
            "width": 10
          },
          {
            "id": "K_1",
            "text": "𐡘",
            "width": 10
          },
          {
            "id": "K_2",
            "text": "𐡙",
            "width": 10
          },
          {
            "id": "K_3",
            "text": "𐡚",
            "width": 10
          },
          {
            "id": "K_4",
            "text": "𐡛",
            "width": 10
          },
          {
            "id": "K_5",
            "text": "𐡜",
            "width": 10
          },
          {
            "id": "K_6",
            "text": "𐡝",
            "width": 10
          },
          {
            "id": "K_7",
            "text": "𐡞",
            "width": 10
          },
          {
            "id": "K_8",
            "text": "𐡟",
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_HYPHEN",
            "text": "-",
            "width": 10
          },
          {
            "id": "K_EQUAL",
            "text": "=",
            "width": 10
          }
        ]
      },
      {
        "id": 1,
        "key": [
          {
            "id": "K_Q",
            "text": "𐡒",
            "width": 10
          },
          {
            "id": "K_W",
            "text": "𐡅",
            "width": 10
          },
          {
            "id": "K_E",
            "text": "𐡄",
            "width": 10
          },
          {
            "id": "K_R",
            "text": "𐡓",
            "width": 10
          },
          {
            "id": "K_T",
            "text": "𐡈",
            "width": 10
          },
          {
            "id": "K_Y",
            "text": "𐡉",
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_I",
            "text": "𐡏",
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_P",
            "text": "𐡐",
            "width": 10
          },
          {
            "id": "K_LBRKT",
            "text": "[",
            "width": 10
          },
          {
            "id": "K_RBRKT",
            "text": "]",
            "width": 10
          },
          {
            "id": "K_BKSLASH",
            "text": "\\",
            "width": 10
          }
        ]
      },
      {
        "id": 2,
        "key": [
          {
            "id": "K_A",
            "text": "𐡀",
            "width": 10
          },
          {
            "id": "K_S",
            "text": "𐡎",
            "width": 10
          },
          {
            "id": "K_D",
            "text": "𐡃",
            "width": 10
          },
          {
            "id": "K_F",
            "text": "𐡕",
            "width": 10
          },
          {
            "id": "K_G",
            "text": "𐡂",
            "width": 10
          },
          {
            "id": "K_H",
            "text": "𐡇",
            "width": 10
          },
          {
            "id": "K_J",
            "text": "𐡑",
            "width": 10
          },
          {
            "id": "K_K",
            "text": "𐡊",
            "width": 10
          },
          {
            "id": "K_L",
            "text": "𐡋",
            "width": 10
          },
          {
            "id": "K_COLON",
            "text": ";",
            "width": 10
          },
          {
            "id": "K_QUOTE",
            "text": "'",
            "width": 10
          }
        ]
      },
      {
        "id": 3,
        "key": [
          {
            "id": "K_Z",
            "text": "𐡆",
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_C",
            "text": "𐡔",
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_B",
            "text": "𐡁",
            "width": 10
          },
          {
            "id": "K_N",
            "text": "𐡍",
            "width": 10
          },
          {
            "id": "K_M",
            "text": "𐡌",
            "width": 10
          },
          {
            "id": "K_COMMA",
            "text": ",",
            "width": 10
          },
          {
            "id": "K_PERIOD",
            "text": ".",
            "width": 10
          },
          {
            "id": "K_SLASH",
            "text": "𐡗",
            "width": 10
          }
        ]
      },
      {
        "id": 4,
        "key": [
          {
            "id": "K_SPACE",
            "text": " ",
            "width": 10,
            "layer": "shift"
          }
        ]
      }
    ]
  },
  {
    "id": "shift",
    "row": [
      {
        "id": 0,
        "key": [
          {
            "id": "K_BKQUOTE",
            "text": "~",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_1",
            "text": "!",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_2",
            "text": "@",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_3",
            "text": "#",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_4",
            "text": "$",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_5",
            "text": "%",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_6",
            "text": "^",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_7",
            "text": "&",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_8",
            "text": "*",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_9",
            "text": "(",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_0",
            "text": ")",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_HYPHEN",
            "text": "_",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_EQUAL",
            "text": "+",
            "width": 10,
            "layer": "shift"
          }
        ]
      },
      {
        "id": 1,
        "key": [
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_LBRKT",
            "text": "{",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_RBRKT",
            "text": "}",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_BKSLASH",
            "text": "|",
            "width": 10,
            "layer": "shift"
          }
        ]
      },
      {
        "id": 2,
        "key": [
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_COLON",
            "text": ":",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_QUOTE",
            "text": "\"",
            "width": 10,
            "layer": "shift"
          }
        ]
      },
      {
        "id": 3,
        "key": [
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "sp": 10,
            "width": 10
          },
          {
            "id": "K_COMMA",
            "text": "<",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_PERIOD",
            "text": ">",
            "width": 10,
            "layer": "shift"
          },
          {
            "id": "K_SLASH",
            "text": "?",
            "width": 10,
            "layer": "shift"
          }
        ]
      },
      {
        "id": 4,
        "key": [
          {
            "id": "K_SPACE",
            "text": " ",
            "width": 10,
            "layer": "shift"
          }
        ]
      }
    ]
  }
]
```

</details>

Known limitations:
- I was unable to test this against a keyboard with longpresses, flicks, and/or multitaps at this time.
- It does not properly handle key IDs for keys without a "standard" vkey mapping.
    - If it maps directly to a `K_` key, we handle that perfectly fine.
    - If it _doesn't_, I currently just leave the LDML id for the key, which is... less than ideal.
- This layout generator does not currently insert frame keys.
    - Note that K_BKSP, K_SHIFT, etc are missing.

### CODE REVIEW

File: web/src/engine/keyboard/src/lazySpan.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/**
+ * A span that does not cut out the substring until it absolutely has to!
+ */
+export class LazySpan {
```
Comment (jahorton) [inline web/src/engine/keyboard/src/lazySpan.ts:4]: From `common/models/wordbreakers`, with alterations (for sub-span extraction).

The original may be found at https://github.com/keymanapp/keyman/blob/72c894bf386d0bc36c48172bc0abcf77057fc338/common/models/wordbreakers/src/main/default/index.ts#L69.

File: web/src/test/auto/headless/engine/keyboard/lazySpan.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  end ??= str.length;
+  let num = 0;
+
+  for(let i = end - 1; i >= start; i--) {
+    let val = str.charCodeAt(i);
+    num = num * SINGLE_CHAR_RANGE + val - ENCODED_NUM_BASE;
+  }
+
+  return num;
+}
```
Comment (jahorton) [inline web/src/test/auto/headless/engine/keyboard/lazySpan.js:24]: Originally from #11088 / `epic/trie-encoding`, but with the endianness flipped and different range / base values.

For a source reference...

https://github.com/keymanapp/keyman/blob/636a3ee05c66a55238700c4a985455088418ece7/common/models/templates/src/trie-compression.ts#L20

File: web/src/test/auto/headless/engine/keyboard/lazySpan.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    const fullSpan = new LazySpan(simpleString, 0, simpleString.length);
+    const quickSpan = fullSpan.substring(4, 9);
+    assert.equal(quickSpan.text, 'quick');
+
+    const latterHalf = new LazySpan(simpleString, simpleString.indexOf('jumped'), simpleString.length);
+    const overSpan = latterHalf.substring(7, 11);
+    assert.equal(overSpan.text, 'over');
+  });
+
+  it('ldml parsing scratchspace', () => {
```
Comment (jahorton) [inline web/src/test/auto/headless/engine/keyboard/lazySpan.js:45]: The prototype KMX+ parse.

File: web/src/test/auto/headless/engine/keyboard/lazySpan.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    // The KMX+ part of the fixture is the 'tail' of its file - once started, it is the remainder.
+    assert.equal(kmxPlusLength + kmxPlusOffset, ldmlFixture.length);
+
+    // Sections
+
+    const sectTableLength = decodeNumber(ldmlFixture, kmxPlusOffset + 4, kmxPlusOffset + 8);
+    const sectTableSpan = new LazySpan(ldmlFixture, kmxPlusOffset, kmxPlusOffset + sectTableLength);
+
+    const sectionMap = new Map();
+    sectionMap.set('sect', sectTableSpan);
```
Comment (jahorton) [inline web/src/test/auto/headless/engine/keyboard/lazySpan.js:58]: We build a map of the primary tables.  Each entry is keyed with the table name and "valued" with a span representing _exactly_ the portion of the file representing that table.

File: web/src/engine/keyboard/src/kmxPlus/convertLayerList.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import TouchLayoutSubKey = TouchLayout.TouchLayoutSubKey;
+import TouchLayoutFlick = TouchLayout.TouchLayoutFlick;
+
+import Codes from '../codes.js';
+import { KeySpec, Keys } from './keys.js';
+import { ButtonClasses } from '../keyboards/defaultLayouts.js';
+
+const keyNames = Object.keys(Codes.keyCodes);
+
+export function convertLayerList(layerList: LayerList, keys: Keys) {
```
Comment (jahorton) [inline web/src/engine/keyboard/src/kmxPlus/convertLayerList.ts:15]: This function converts one layer list (one pre-parsed entry of the `layr.lists` subtable) into the OSK-friendly `TouchLayout` format.

File: web/src/engine/keyboard/src/kmxPlus/layr.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+      layerLists.push({
+        name, minDeviceWidth, layers: listLayers
+      });
+    }
+
+    this.layerLists = layerLists;
+  }
+
+  private processLayers(layerTableOffset: number, strs: Strs, keys: Keys) {
```
Comment (jahorton) [inline web/src/engine/keyboard/src/kmxPlus/layr.ts:56]: This method may include some "heavy lifting" on behalf of the `convertLayerList` method.

File: web/src/engine/keyboard/src/kmxPlus/ldmlKeyboardObject.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    if(this.sect.name != 'sect') {
+      throw new Error("Provided KMX+ data did not start with the `sect` table!")
+    }
+
+    /*
+     * Preferred order:
+     * - strs
+     * - list - needs strs
+     * - keys - needs lsts
+     * - layr - needs keys, others.
```
Comment (jahorton) [inline web/src/engine/keyboard/src/kmxPlus/ldmlKeyboardObject.ts:43]: `disp` could probably go after `strs` or `list` - it only really needs to reference `strs`, I believe.

File: web/src/engine/keyboard/src/kmxPlus/utils.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  return num;
+}
+
+export function decodeSectName(data: Uint8Array, start: number) {
+  return [...data.slice(start, start+4)].map((x) => String.fromCharCode(x)).join('');
+}
+
+export function decodeString(data: Uint8Array, start: number, len: number) {
```
Comment (jahorton) [inline web/src/engine/keyboard/src/kmxPlus/utils.ts:21]: For the UTF-16LE `strs` table.

File: web/src/test/auto/headless/engine/keyboard/kmx-plus-parsing.spec.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  });
+
+  it("correctly builds a touch-layout from an LDML layer spec", () => {
+    const kmxPlusOffset = ldmlFixtureAsAscii.indexOf('sect');
+    const kbdObj = new LdmlKeyboardObject(ldmlFixture, kmxPlusOffset);
+
+    const layout = convertLayerList(kbdObj.layr.layerLists[0], kbdObj.keys);
+
+    console.log(JSON.stringify(layout, null, 2));
```
Comment (jahorton) [inline web/src/test/auto/headless/engine/keyboard/kmx-plus-parsing.spec.js:34]: This line is the source of the example touch-layout I placed in the PR's description.

File: docs/file-formats/kmx-plus-file-format.md [hunk 10]
```diff
@@ -173,8 +173,8 @@ Then for each string:
 
 | ∆ | Bits | Name          | Description                                   |
 |---|------|---------------|-----------------------------------------------|
-|16+|  32  | offset        | off: Offset to string                         |
-|20+|  32  | length        | int: Length of string in UTF-16LE code units  |
+|12+|  32  | offset        | off: Offset to string                         |
```
Comment (srl295) [inline docs/file-formats/kmx-plus-file-format.md:176]: oops, good catch!

File: web/src/engine/keyboard/src/kmxPlus/keys.ts [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+      if(i == 0) {
+        if(elemCount == 0 && firstIndex == 0 && stringIndex == 0) {
+          continue;
+        } else {
+          throw new Error("Empty flick set expected at index 0 in `keys.flicks` subtable");
+        }
+      }
+
+      // TODO:  no good current fixture available with data for this.
```
Comment (srl295) [inline web/src/engine/keyboard/src/kmxPlus/keys.ts:86]: unit tests only

File: web/src/engine/keyboard/src/kmxPlus/keys.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    this.keys = keyMap;
+  }
+
+  private processFlickSubtables(strs: Strs, list: List, baseOffset: number) {
+    const rawData = this.data;
+
+    const keyCount    = decodeNumber(rawData,  8);
+    const flicksCount = decodeNumber(rawData, 12);
+
+    const FLICK_GROUP_LEN = 12;
```
Comment (srl295) [inline web/src/engine/keyboard/src/kmxPlus/keys.ts:62]: now, shouldn't this file be able to use https://github.com/keymanapp/keyman/blob/master/core/include/ldml/keyman_core_ldml.ts ? 
so:

```suggestion
    const FLICK_GROUP_LEN = Constants.length_keys_flick_list;
```

File: web/src/engine/keyboard/src/kmxPlus/ldmlKeyboardObject.ts (hunk 7 above)
Comment (srl295) [reply web/src/engine/keyboard/src/kmxPlus/ldmlKeyboardObject.ts:43]: correct

**Review by srl295 — APPROVED** [review]
reasonable though should be able to reuse constants from the compiler side?

**Comment by mcdurdin** [comment]
We are approaching this from a different direction and keeping kmx+ interpretation entirely within Core

**Comment by srl295** [comment]
> We are approaching this from a different direction and keeping kmx+ interpretation entirely within Core

without sharing constants?

**Comment by mcdurdin** [comment]
> without sharing constants?

A few constants are helpful, e.g. VKs and modifiers but those are not kmx+-specific.

**Comment by srl295** [comment]
> > without sharing constants?
> 
> A few constants are helpful, e.g. VKs and modifiers but those are not kmx+-specific.

I think you are saying "a different direction _from this PR_". - my comment above is that https://github.com/keymanapp/keyman/blob/master/core/include/ldml/keyman_core_ldml.ts contains the offsets and KMX+ specific constants from the spec and could be shared here.

Speaking of spec, perhaps the spec corrections should be split out from this PR?

**Comment by mcdurdin** [comment]
> Speaking of spec, perhaps the spec corrections should be split out from this PR?

Good point, will do.

**Comment by mcdurdin** [comment]
Closing this. We will parse kmx+ layout data in Keyman Core and pass it to Keyman Engine for Web through a new Core API.

**Pull request closed without merging**